File: python/200_number_of_islands.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        island_count = 0

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    self.findIsland(i, j, grid)
                    island_count += 1
        
        return island_count

    def findIsland(self, i, j, grid):
        """
        :type i, j: int
        :type grid: List[List[str]]
        :rtype: None
        """
        queue = [(i, j)]
        x_max = len(grid) - 1
        y_max = len(grid[0]) - 1

        while len(queue) != 0:
            grid_square = queue.pop()
            gx = grid_square[0]
            gy = grid_square[1]
            if grid[gx][gy] == '1':
                grid[gx][gy] = '2'
                if gx > 0 and grid[gx - 1][gy] == '1':
                    queue.append((gx - 1, gy))
                if gx < x_max and grid[gx + 1][gy] == '1':
                    queue.append((gx + 1, gy))
                if gy > 0 and grid[gx][gy - 1] == '1':
                    queue.append((gx, gy - 1))
                if gy < y_max and grid[gx][gy + 1] == '1':
                    queue.append((gx, gy + 1))
            

File: c++/200_number_of_islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        queue<pair<int, int>> to_visit;
        int island_count = 0;

        // Iterate over grid, checking all coord pairs for 'land,'
        // if found, perform BFS starting at that coord to remove
        // all 'land' attached to that 'island,' and continue
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] == '1') {
                    to_visit.push(make_pair(i, j));
                    grid[i][j] = '0';
                    while (!to_visit.empty()) {
                        int x, y;
                        tie(x, y) = to_visit.front();
                        to_visit.pop();

                        // Enqueue all neighboring island, and zero out that land
                        // to prevent requeueing of already queued land
                        if (x > 0 && grid[x-1][y] == '1') {
                            to_visit.push(make_pair(x-1, y));
                            grid[x-1][y] = '0';
                        }
                        if (x < grid.size() - 1 && grid[x+1][y] == '1') {
                            to_visit.push(make_pair(x+1, y));
                            grid[x+1][y] = '0';
                        }
                        if (y > 0 && grid[x][y-1] == '1') {
                            to_visit.push(make_pair(x, y-1));
                            grid[x][y-1] = '0';
                        }
                        if (y < grid[0].size() - 1 && grid[x][y+1] == '1') {
                            to_visit.push(make_pair(x, y+1));
                            grid[x][y+1] = '0';
                        }
                    }
                    island_count += 1;
                }
            }
        }

        return island_count;
    }
};
